### PR TITLE
Adds DFS, BFS and Dijkstra's algorithm solutions for "The Maze II"

### LINKED ISSUE #56: 505. The Maze II
https://github.com/altayhunoglu/algorithms/issues/48 sorusu ile aynı.
Ama bu sefer, hedefe varmanın mümkün olup olmadığını değil, hedefe giden en kısa mesafeyi ölçmemizi istiyor.

LeetCode: https://leetcode.com/problems/the-maze-ii

Varsayılan kod:

```
class Solution {
    public int shortestDistance(int[][] maze, int[] start, int[] destination) {
        
    }
}
```

image
image
image
image

### PR DESCRIPTION
**Pull request by altayhunoglu**
resolves #56

This question is a version of https://github.com/spiralgo/algorithms/pull/282, but this time we are asked to find the shortest path to the target.

Dijkstra's algorithm appears to be the fastest with ~5ms and the way it searches is almost the same as the BFS.

**But Dijkstra starts with detecting the node which has the least distance from the starting point.**

`Priority Queue` is a perfect match to poll the closest node.

### CODE REVIEW

**Comment by ErdemT09** [comment]
Issue: #56

**Comment by altayhunoglu** [comment]
Hello @ErdemT09,
 
I have done some improvements in Dijkstra implementation.
The latest commit:
- removes `distance` array as it is obsolete after we implement PQ to get the closest node.
- removes `visited` array to improve space efficiency, utilizes the `maze` matrix for this purpose.
- removes unnecessary distance compare as PQ already does this.

File: src/main/java/algorithms/curated170/medium/maze2/TheMazeIIDijkstra.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+           
+            if (startPoint[0] < 0)
+                break;
+            
+            int x = startPoint[0];
+            int y = startPoint[1];
+            
+            if(maze[x][y]==-1) continue;
+            
+            maze[x][y] = -1;
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/maze2/TheMazeIIDijkstra.java:35]: I reccommend moving -1 to a constant called "VISITED" for readability.

File: src/main/java/algorithms/curated170/medium/maze2/TheMazeIIDijkstra.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        PriorityQueue<int[]> pq = new PriorityQueue<>((p1, p2) -> p1[2] - p2[2]); 
+        int[][] directions = {{0, -1}, {1, 0}, {-1, 0}, {0, 1}};
+     
+        pq.add(new int[]{start[0], start[1], 0});
+        while (!pq.isEmpty()) {
+        
+            int[] startPoint = pq.poll();
+           
+            if (startPoint[0] < 0)
+                break;
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/maze2/TheMazeIIDijkstra.java:28]: I don't understand the function of this statement here. Can you please explain that?

File: src/main/java/algorithms/curated170/medium/maze2/TheMazeIIDijkstra.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            }
+        }
+        return -1;
+    }
+ 
+     boolean canPass(int nextX, int nextY) {
+            return (nextX >= 0 && nextY >= 0
+                    && nextX < maze.length
+                    && nextY < maze[0].length
+                    && maze[nextX][nextY] != 1);
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/maze2/TheMazeIIDijkstra.java:74]: Why don't we check for this condition unlike in the previous two solutions?
`maze[nextX][nextY] == 0);`

**Comment by ErdemT09** [comment]
I have accidentally deleted the branch. I should have been more careful.

File: src/main/java/algorithms/curated170/medium/maze2/TheMazeIIDijkstra.java (hunk 3 above)
Comment (ErdemT09) [reply src/main/java/algorithms/curated170/medium/maze2/TheMazeIIDijkstra.java:71]: I have understood the reason why we don't do this.
In the previous solution, the tiles that we come across are either empty or a wall. If empty, we move through them. If wall, we stop. Here, we have a three valued logic: A tile is either empty but visited, empty or a wall. We should be able to move through an empty tile even if it is visited. At the end, we can't stop and branch off there, but whilst moving to another valid tile, we should be able to pass them.

File: src/main/java/algorithms/curated170/medium/maze2/TheMazeIIDijkstra.java (hunk 1 above)
Comment (ErdemT09) [reply src/main/java/algorithms/curated170/medium/maze2/TheMazeIIDijkstra.java:32]: I still suggest doing this.

File: src/main/java/algorithms/curated170/medium/maze2/TheMazeIIDijkstra.java (hunk 3 above)
Comment (altayhunoglu) [reply src/main/java/algorithms/curated170/medium/maze2/TheMazeIIDijkstra.java:71]: Thanks for the explanation Erdem. Appreciated.

File: src/main/java/algorithms/curated170/medium/maze2/TheMazeIIDijkstra.java (hunk 1 above)
Comment (altayhunoglu) [reply src/main/java/algorithms/curated170/medium/maze2/TheMazeIIDijkstra.java]: Oke, it is done. Thanks for the suggestion.

**Review by ErdemT09 — APPROVED** [review]
Approved. A good question in terms of graph theory.

**Pull request merged**